Mathlib/NumberTheory/NumberField/CanonicalEmbedding/FundamentalCone.lean

```diff
@@ -234,6 +237,100 @@ theorem unit_smul_mem_iff_mem_torsion (hx : x ∈ fundamentalCone K) (u : (𝓞
   · rw [AddSubmonoid.mk_vadd, vadd_eq_add, zero_add]
     exact hx.1
 
+variable (K) in
+/-- The set of images by `mixedEmbedding` of algebraic integers of `K` contained in the
+fundamental cone. -/
+def integralPoint : Set (mixedSpace K) :=
+  fundamentalCone K ∩ (mixedEmbedding.integerLattice K)
+
+theorem mem_integralPoint {a : mixedSpace K} :
+    a ∈ integralPoint K ↔ a ∈ fundamentalCone K ∧ ∃ x : (𝓞 K), mixedEmbedding K x = a:= by
+  simp only [integralPoint, Set.mem_inter_iff, SetLike.mem_coe, LinearMap.mem_range,
+    AlgHom.toLinearMap_apply, RingHom.toIntAlgHom_coe, RingHom.coe_comp, Function.comp_apply]
+
+/-- If `a` is an integral point, then there is a *unique* algebraic integer in `𝓞 K` such
+that `mixedEmbedding K x = a`. -/
+theorem exists_unique_preimage_of_integralPoint {a : mixedSpace K} (ha : a ∈ integralPoint K) :
+    ∃! x : (𝓞 K), mixedEmbedding K x = a := by
+  obtain ⟨_, ⟨x, rfl⟩⟩ := mem_integralPoint.mp ha
+  refine Function.Injective.exists_unique_of_mem_range ?_ (Set.mem_range_self x)
+  exact (mixedEmbedding_injective K).comp RingOfIntegers.coe_injective
+
+theorem integralPoint_ne_zero (a : integralPoint K) :
+    (a : mixedSpace K) ≠ 0 := by
+  by_contra!
+  exact a.prop.1.2 (this.symm ▸ mixedEmbedding.norm.map_zero')
+
+open scoped nonZeroDivisors
+
+/-- For `a : integralPoint K`, the unique nonzero algebraic integer whose image by
+`mixedEmbedding` is equal to `a`, see `mixedEmbedding_preimageOfIntegralPoint`. -/
+def preimageOfIntegralPoint (a : integralPoint K) : (𝓞 K)⁰ :=
+  ⟨(mem_integralPoint.mp a.prop).2.choose, mem_nonZeroDivisors_of_ne_zero (by
+  simp_rw [ne_eq, ← RingOfIntegers.coe_injective.eq_iff, ← (mixedEmbedding_injective K).eq_iff,
+    map_zero, (mem_integralPoint.mp a.prop).2.choose_spec, integralPoint_ne_zero,
+    not_false_eq_true])⟩
+
+@[simp]
+theorem mixedEmbedding_preimageOfIntegralPoint (a : integralPoint K) :
+    mixedEmbedding K (preimageOfIntegralPoint a : 𝓞 K) = (a : mixedSpace K) := by
+  rw [preimageOfIntegralPoint, (mem_integralPoint.mp a.prop).2.choose_spec]
+
+theorem preimageOfIntegralPoint_mixedEmbedding {x : (𝓞 K)⁰}
```

Comment:
Is there a reason why you work with nonzero divisors instead of nonzero elements?

Reply:
Yes, I want to use the general isomorphism `Ideal.associatesNonZeroDivisorsEquivIsPrincipal`.

Reply:
OK, can you explain this in the docstring? Thanks!

Reply:
I'll do it the next PR then so I can point to the result where the isomorphism is used.